File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/qs/CustomQsRoundCorner.java
package com.qimian233.ztool.hook.modules.systemui.qs;

import android.annotation.SuppressLint;
import android.graphics.drawable.ClipDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.LayerDrawable;
import android.graphics.drawable.RippleDrawable;
import android.graphics.drawable.StateListDrawable;
import android.widget.ProgressBar;

import com.qimian233.ztool.data.keys.ScopeKeys;
import com.qimian233.ztool.data.keys.PreferenceKeys;
import com.qimian233.ztool.hook.base.AppHookModule;

import io.github.libxposed.api.XposedModuleInterface;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;

@SuppressLint("PrivateApi")
public class CustomQsRoundCorner extends AppHookModule {

    private static int headUpTileRoundCornerRadius = 32;
    private static int normalTileRoundCornerRadius = 96;

    public CustomQsRoundCorner() {}

    @Override
    public String getModuleName() {
        return "qs_round_corner";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[] {ScopeKeys.SYSTEM_UI.packageName};
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();
        updateRoundCornerPrefs();

        // Head-up tiles
        Method changeCornerRadiusMethod = classLoader
                .loadClass("com.android.systemui.qs.tileimpl.QSTileViewImpl")
                .getDeclaredMethod("changeCornerRadius", float.class);
        hookWithId(changeCornerRadiusMethod, "change_corner_radius", chain -> chain.proceed(new Object[]{(float) headUpTileRoundCornerRadius}));

        // Normal tiles
        Method updateRippleRadiusMethod = classLoader
                .loadClass("com.android.systemui.qs.tileimpl.CustomQSTileViewImpl")
                .getDeclaredMethod("updateRippleRadius");
        hookWithId(updateRippleRadiusMethod, "update_ripple_radius", chain -> {
            Object result = chain.proceed();
            try {
                Class<?> cl = chain.getThisObject().getClass();

                RippleDrawable rippleDrawable =
                        (RippleDrawable) findField(cl, "qsTileBackground").get(chain.getThisObject());

                if (rippleDrawable != null) {
                    Drawable mask = rippleDrawable.findDrawableByLayerId(android.R.id.mask);
                    if (mask instanceof GradientDrawable) {
                        ((GradientDrawable) mask).setCornerRadius((float) normalTileRoundCornerRadius);
                    }
                }

                LayerDrawable backgroundDrawable =
                        (LayerDrawable) findField(cl, "backgroundDrawable").get(chain.getThisObject());

                if (backgroundDrawable != null) {
                    int count = backgroundDrawable.getNumberOfLayers();
                    for (int i = 0; i < count; i++) {
                        Drawable layer = backgroundDrawable.getDrawable(i);
                        if (layer instanceof GradientDrawable) {
                            ((GradientDrawable) layer).setCornerRadius((float) normalTileRoundCornerRadius);
                        }
                    }
                }
            } catch (Exception e) {
                logger.error("Cannot set normal tile round corner radius!", e);
            }
            return result;
        });

        // Sliders
        Class<?> toggleSliderViewClass = classLoader.loadClass("com.android.systemui.settings.ToggleSliderView");

        Method refreshSeekBarMethod = toggleSliderViewClass.getDeclaredMethod("refreshSeekBar", ProgressBar.class);
        hookWithId(refreshSeekBarMethod, "refresh_seek_bar", chain -> {
            logger.debug("refreshSeekBar afterHookedMethod called!");
            Object result = chain.proceed();
            applySeekBarRoundCorner((ProgressBar) chain.getArg(0));
            return result;
        });

        Method updateBrightnessSliderMethod = toggleSliderViewClass.getDeclaredMethod("updateBrightnessSlider");
        hookWithId(updateBrightnessSliderMethod, "update_brightness_slider", chain -> {
            Object result = chain.proceed();
            logger.debug("updateBrightnessSlider afterHookedMethod called!");
            Class<?> cl = chain.getThisObject().getClass();
            ProgressBar brightnessSlider = (ProgressBar) cl.getDeclaredField("mBrightnessSlider")
                    .get(chain.getThisObject());
            if (brightnessSlider != null) {
                cl.getDeclaredMethod("refreshSeekBar", ProgressBar.class)
                        .invoke(chain.getThisObject(), brightnessSlider);
            }
            return result;
        });

        Method updateVolumeSliderMethod = toggleSliderViewClass.getDeclaredMethod("updateVolumeSlider");
        hookWithId(updateVolumeSliderMethod, "update_volume_slider", chain -> {
            Object result = chain.proceed();
            logger.debug("updateVolumeSlider afterHookedMethod called!");
            Class<?> cl = chain.getThisObject().getClass();
            ProgressBar mediaSlider = (ProgressBar) cl.getDeclaredField("mMediaVolumeSlider")
                    .get(chain.getThisObject());
            if (mediaSlider != null) {
                cl.getDeclaredMethod("refreshSeekBar", ProgressBar.class)
                        .invoke(chain.getThisObject(), mediaSlider);
            }
            return result;
        });

        Constructor<?> toggleCtor = toggleSliderViewClass.getDeclaredConstructor(
                android.content.Context.class,
                android.util.AttributeSet.class,
                int.class);
        hookWithId(toggleCtor, "toggle", chain -> {
            chain.proceed();
            Class<?> cl = chain.getThisObject().getClass();
            ProgressBar brightnessSlider = (ProgressBar) cl.getDeclaredField("mBrightnessSlider")
                    .get(chain.getThisObject());
            ProgressBar mediaSlider = (ProgressBar) cl.getDeclaredField("mMediaVolumeSlider")
                    .get(chain.getThisObject());
            if (brightnessSlider != null) {
                cl.getDeclaredMethod("refreshSeekBar", ProgressBar.class)
                        .invoke(chain.getThisObject(), brightnessSlider);
            }
            if (mediaSlider != null) {
                cl.getDeclaredMethod("refreshSeekBar", ProgressBar.class)
                        .invoke(chain.getThisObject(), mediaSlider);
            }
            return null;
        });
    }

    private void applySeekBarRoundCorner(ProgressBar progressBar) {
        if (progressBar == null) {
            logger.debug("applySeekBarRoundCorner skipped from " + "refreshSeekBar" + ": progressBar is null");
            return;
        }

        Drawable progressDrawable = progressBar.getProgressDrawable();
        if (!(progressDrawable instanceof LayerDrawable)) {
            logger.debug("applySeekBarRoundCorner skipped from " + "refreshSeekBar" + ": progress drawable is " + describeDrawable(progressDrawable));
            return;
        }

        try {
            LayerDrawable layerDrawable = (LayerDrawable) progressDrawable;
            Drawable backgroundDrawable = layerDrawable.findDrawableByLayerId(android.R.id.background);
            if (backgroundDrawable instanceof GradientDrawable) {
                ((GradientDrawable) backgroundDrawable).setCornerRadius((float) headUpTileRoundCornerRadius);
            } else {
                logger.debug("Background layer is " + describeDrawable(backgroundDrawable) + " from " + "refreshSeekBar");
            }

            Drawable progressLayer = layerDrawable.findDrawableByLayerId(android.R.id.progress);
            if (progressLayer instanceof StateListDrawable) {
                StateListDrawable stateListDrawable = (StateListDrawable) progressLayer;
                for (int i = 0; i < stateListDrawable.getStateCount(); i++) {
                    Drawable stateDrawable = stateListDrawable.getStateDrawable(i);
                    if (stateDrawable instanceof ClipDrawable) {
                        Drawable innerDrawable = ((ClipDrawable) stateDrawable).getDrawable();
                        if (innerDrawable instanceof GradientDrawable) {
                            ((GradientDrawable) innerDrawable).setCornerRadius((float) headUpTileRoundCornerRadius);
                        }
                    }
                }
            } else {
                logger.debug("Progress layer is " + describeDrawable(progressLayer) + " from " + "refreshSeekBar");
            }
        } catch (Throwable t) {
            logger.error("applySeekBarRoundCorner failed from " + "refreshSeekBar", t);
        }
    }

    private String describeDrawable(Drawable drawable) {
        return drawable == null ? "null" : drawable.getClass().getName();
    }

    private void updateRoundCornerPrefs() {
        try {
            headUpTileRoundCornerRadius = getRemotePreferences().getInt(PreferenceKeys.HEAD_UP_ROUND_CORNER_RADIUS.name, 32);
        } catch (Throwable t) {
            headUpTileRoundCornerRadius = 32;
        }
        try {
            normalTileRoundCornerRadius = getRemotePreferences().getInt(PreferenceKeys.TILE_ROUND_CORNER_RADIUS.name, 96);
        } catch (Throwable t) {
            normalTileRoundCornerRadius = 96;
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/systemui/qs/CustomQsRoundCorner.kt
package com.qimian233.ztool.hook.modules.systemui.qs

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.drawable.ClipDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.LayerDrawable
import android.graphics.drawable.RippleDrawable
import android.graphics.drawable.StateListDrawable
import android.util.AttributeSet
import android.widget.ProgressBar
import com.qimian233.ztool.data.keys.PreferenceKeys
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.hook.base.AppHookModule
import io.github.libxposed.api.XposedModuleInterface.PackageLoadedParam
import java.lang.reflect.Constructor
import java.lang.reflect.Method

@SuppressLint("PrivateApi")
class CustomQsRoundCorner : AppHookModule() {

    private var headUpTileRoundCornerRadius = 32
    private var normalTileRoundCornerRadius = 96

    override fun getModuleName(): String = PreferenceKeys.QS_ROUND_CORNER.name

    override fun getTargetPackages(): Array<String> = arrayOf(ScopeKeys.SYSTEM_UI.packageName)

    override fun handleLoadPackage(param: PackageLoadedParam) {
        val classLoader = param.defaultClassLoader
        updateRoundCornerPrefs()

        // Head-up tiles
        val changeCornerRadiusMethod: Method = classLoader
            .loadClass("com.android.systemui.qs.tileimpl.QSTileViewImpl")
            .getDeclaredMethod("changeCornerRadius", Float::class.javaPrimitiveType)
        hookWithId(changeCornerRadiusMethod, "change_corner_radius") { chain ->
            chain.proceed(arrayOf(headUpTileRoundCornerRadius.toFloat()))
        }

        // Normal tiles
        val updateRippleRadiusMethod: Method = classLoader
            .loadClass("com.android.systemui.qs.tileimpl.CustomQSTileViewImpl")
            .getDeclaredMethod("updateRippleRadius")
        hookWithId(updateRippleRadiusMethod, "update_ripple_radius") { chain ->
            val result = chain.proceed()
            try {
                val cl = chain.thisObject.javaClass

                val rippleDrawable =
                    findField(cl, "qsTileBackground").get(chain.thisObject) as? RippleDrawable

                if (rippleDrawable != null) {
                    val mask = rippleDrawable.findDrawableByLayerId(android.R.id.mask)
                    if (mask is GradientDrawable) {
                        mask.cornerRadius = normalTileRoundCornerRadius.toFloat()
                    }
                }

                val backgroundDrawable =
                    findField(cl, "backgroundDrawable").get(chain.thisObject) as? LayerDrawable

                if (backgroundDrawable != null) {
                    val count = backgroundDrawable.numberOfLayers
                    for (i in 0 until count) {
                        val layer = backgroundDrawable.getDrawable(i)
                        if (layer is GradientDrawable) {
                            layer.cornerRadius = normalTileRoundCornerRadius.toFloat()
                        }
                    }
                }
            } catch (e: Exception) {
                logger.error("Cannot set normal tile round corner radius!", e)
            }
            result
        }

        // Sliders
        val toggleSliderViewClass = classLoader.loadClass("com.android.systemui.settings.ToggleSliderView")

        val refreshSeekBarMethod: Method =
            toggleSliderViewClass.getDeclaredMethod("refreshSeekBar", ProgressBar::class.java)
        hookWithId(refreshSeekBarMethod, "refresh_seek_bar") { chain ->
            logger.debug("refreshSeekBar afterHookedMethod called!")
            val result = chain.proceed()
            applySeekBarRoundCorner(chain.args[0] as? ProgressBar)
            result
        }

        val updateBrightnessSliderMethod: Method =
            toggleSliderViewClass.getDeclaredMethod("updateBrightnessSlider")
        hookWithId(updateBrightnessSliderMethod, "update_brightness_slider") { chain ->
            val result = chain.proceed()
            logger.debug("updateBrightnessSlider afterHookedMethod called!")
            val cl = chain.thisObject.javaClass
            val brightnessSlider =
                cl.getDeclaredField("mBrightnessSlider").get(chain.thisObject) as? ProgressBar
            if (brightnessSlider != null) {
                cl.getDeclaredMethod("refreshSeekBar", ProgressBar::class.java)
                    .invoke(chain.thisObject, brightnessSlider)
            }
            result
        }

        val updateVolumeSliderMethod: Method =
            toggleSliderViewClass.getDeclaredMethod("updateVolumeSlider")
        hookWithId(updateVolumeSliderMethod, "update_volume_slider") { chain ->
            val result = chain.proceed()
            logger.debug("updateVolumeSlider afterHookedMethod called!")
            val cl = chain.thisObject.javaClass
            val mediaSlider =
                cl.getDeclaredField("mMediaVolumeSlider").get(chain.thisObject) as? ProgressBar
            if (mediaSlider != null) {
                cl.getDeclaredMethod("refreshSeekBar", ProgressBar::class.java)
                    .invoke(chain.thisObject, mediaSlider)
            }
            result
        }

        val toggleCtor: Constructor<*> = toggleSliderViewClass.getDeclaredConstructor(
            Context::class.java,
            AttributeSet::class.java,
            Int::class.javaPrimitiveType
        )
        hookWithId(toggleCtor, "toggle") { chain ->
            chain.proceed()
            val cl = chain.thisObject.javaClass
            val brightnessSlider =
                cl.getDeclaredField("mBrightnessSlider").get(chain.thisObject) as? ProgressBar
            val mediaSlider =
                cl.getDeclaredField("mMediaVolumeSlider").get(chain.thisObject) as? ProgressBar
            if (brightnessSlider != null) {
                cl.getDeclaredMethod("refreshSeekBar", ProgressBar::class.java)
                    .invoke(chain.thisObject, brightnessSlider)
            }
            if (mediaSlider != null) {
                cl.getDeclaredMethod("refreshSeekBar", ProgressBar::class.java)
                    .invoke(chain.thisObject, mediaSlider)
            }
            null
        }
    }

    private fun applySeekBarRoundCorner(progressBar: ProgressBar?) {
        if (progressBar == null) {
            logger.debug("applySeekBarRoundCorner skipped from refreshSeekBar: progressBar is null")
            return
        }

        val progressDrawable = progressBar.progressDrawable
        if (progressDrawable !is LayerDrawable) {
            logger.debug(
                "applySeekBarRoundCorner skipped from refreshSeekBar: progress drawable is "
                    + describeDrawable(progressDrawable)
            )
            return
        }

        try {
            val backgroundDrawable = progressDrawable.findDrawableByLayerId(android.R.id.background)
            if (backgroundDrawable is GradientDrawable) {
                backgroundDrawable.cornerRadius = headUpTileRoundCornerRadius.toFloat()
            } else {
                logger.debug(
                    "Background layer is " + describeDrawable(backgroundDrawable) + " from refreshSeekBar"
                )
            }

            val progressLayer = progressDrawable.findDrawableByLayerId(android.R.id.progress)
            if (progressLayer is StateListDrawable) {
                for (i in 0 until progressLayer.stateCount) {
                    val stateDrawable = progressLayer.getStateDrawable(i)
                    if (stateDrawable is ClipDrawable) {
                        val innerDrawable = stateDrawable.drawable
                        if (innerDrawable is GradientDrawable) {
                            innerDrawable.cornerRadius = headUpTileRoundCornerRadius.toFloat()
                        }
                    }
                }
            } else {
                logger.debug(
                    "Progress layer is " + describeDrawable(progressLayer) + " from refreshSeekBar"
                )
            }
        } catch (t: Throwable) {
            logger.error("applySeekBarRoundCorner failed from refreshSeekBar", t)
        }
    }

    private fun describeDrawable(drawable: Drawable?): String {
        return drawable?.javaClass?.name ?: "null"
    }

    private fun updateRoundCornerPrefs() {
        headUpTileRoundCornerRadius = try {
            remotePreferences.getInt(PreferenceKeys.HEAD_UP_ROUND_CORNER_RADIUS.name, 32)
        } catch (_: Throwable) {
            32
        }
        normalTileRoundCornerRadius = try {
            remotePreferences.getInt(PreferenceKeys.TILE_ROUND_CORNER_RADIUS.name, 96)
        } catch (_: Throwable) {
            96
        }
    }
}
